### DISCUSSION TITLE
Closing popup window leaves border as orphan

### DISCUSSION THREAD

**Discussion by cseickel** [MunifTanjim/nui.nvim#99] [General]
I've noticed a bug with the popup window that happens consistently. If the window is closed in any way other than calling the unmount() method, the inner window is destroyed but the outer border window remains and it is (nearly) impossible for the end user to close it.

Examples of what cause this would be executing `:q` or `CTRL+w q` in the popup window.

I tried it with the example at https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/popup and had the same result.

I would think that setting this event handler would handle the window close situation,  but it does not:

```lua
local event = require("nui.utils.autocmd").event
popup:on({ event.BufLeave, event.BufDelete }, function()
    popup:unmount()
end, { once = true })
```

Any ideas on how to force unmount to happen when the inner window closes?

**Comment by cseickel** [comment]
Sometimes going through the trouble of explaining a problem helps to figure it out on your own. Setting `bufoptions.bufhidden = "delete"` makes it work as expected. That definitely works for my use case. It may not work for everyone though.

**Comment by MunifTanjim** [comment]
```lua
local function test_98()
  -- example from  https://github.com/MunifTanjim/nui.nvim/tree/main/lua/nui/popup
  local popup = Popup({
    position = "50%",
    size = {
      width = 80,
      height = 40,
    },
    enter = true,
    focusable = true,
    zindex = 50,
    relative = "editor",
    border = {
      padding = {
        top = 2,
        bottom = 2,
        left = 3,
        right = 3,
      },
      style = "rounded",
      text = {
        top = " I am top title ",
        top_align = "center",
        bottom = "I am bottom title",
        bottom_align = "left",
      },
    },
    buf_options = {
      modifiable = true,
      readonly = false,
    },
    win_options = {
      winblend = 10,
      winhighlight = "Normal:Normal,FloatBorder:FloatBorder",
    },
  })

  popup:mount()

  popup:on({ event.BufLeave, event.BufDelete }, function()
    popup:unmount()
  end, { once = true })
end
```

I can't reproduce the issue with the code above 🤔

Doing both `:q` and `CTRL+w q` in the popup window closes both popup and border.

**Comment by MunifTanjim** [comment]
Which neovim version are you using?

I'm on latest head: `NVIM v0.7.0-dev+859-g5623b3c82`

**Comment by cseickel** [comment]
Thanks for the quick reply!

Strange, I swear my example was not working before but it is working now. 🤷

One thing that is still an issue is that it is the BufLeave event that is actually closing the window in this example, not BufDelete. If you want to allow the window to be unfocused, say because you want a popup over your popup, then you can't use BufLeave and there needs to be another solution to prevent creating an orphaned border window.

**Comment by MunifTanjim** [comment]
Can you try `BufWinLeave` or `BufHidden` autocmd?

```lua
  popup:on({ event.BufWinLeave }, function()
    vim.schedule(function()
      popup:unmount()
    end)
  end, { once = true })
```

**Comment by MunifTanjim** [comment]
The can be many variants of this scenario... it depends on what you're trying to do and what buffer options you've set. Just need to choose the right solution for the situation at hand 😄 

I hope this helps.

**Comment by cseickel** [comment]
BufWinLeave almost works, but I get this error:
```
Error detected while processing BufWinLeave Autocommands for "<buffer=66>":
E937: Attempt to delete a buffer that is in use
```
BufHidden does the same thing.

However, using schedule_wrap seems to fix that issue:
```lua
  popup:on({ event.BufHidden }, function()
      vim.schedule_wrap(function ()
        popup:unmount()
      end)()
  end, { once = true })
```

I don't know if I am just abusing that function like a javascript developer uses setTimeout() though!

If this is a correct solution, maybe  there should be a simple way to declare this behavior, or add it to the documentation?

**Comment by MunifTanjim** [comment]
You shouldn't need to use `schedule_wrap` and then call it again 🤔 Just wrapping the callback with `schedule` should work:

```lua
  vim.schedule(function()
    popup:unmount()
  end)
```

I think in this case, this is the right solution. Using `vim.schedule` here means "wait for whatever vim is doing and when that's done run `popup:unmount()`".

**Comment by cseickel** [comment]
Your're right. I'm moving so quickly I didn't even notice you added the vim.schedule() to your code and I was just changing the event names when I tested. Sorry for the confusion, your suggested code works perfectly.

**Comment by MunifTanjim** [comment]
> If this is a correct solution, maybe there should be a simple way to declare this behavior

Like I said, there can be other variations of this situation. So we can't do this by default inside `nui.nvim`. User of `nui.nvim` needs to do what's right for their particular situation.

Well, once you know which autocmd event to use, it's pretty simple already.

> or add it to the documentation?

You can add this to Wiki if you want: https://github.com/MunifTanjim/nui.nvim/wiki/nui.popup

I think it would be helpful for others 😃

---

Closing this issue since it's solved.

**Comment by MunifTanjim** [comment]
Turns out vim has a `QuitPre` autocmd event. After https://github.com/MunifTanjim/nui.nvim/pull/183 is merged, components would automatically call `:unmount` method if `quit` command is run.
